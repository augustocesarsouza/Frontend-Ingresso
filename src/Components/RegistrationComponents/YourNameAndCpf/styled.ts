import styled from 'styled-components';

export const ContainerInputExternal = styled.div`
  display: flex;
  gap: 20px;
`

export const ContainerYourNameAndWarning = styled.div`
  display: flex;
  flex-direction: column;
  width: 55%;
`

interface ContainerAllColorsProps {
  $focusorblur: string;
  $greaterthan3lessthan61: string;
}

export const ContainerAllColors = styled.div<ContainerAllColorsProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  border-width: 1px 1px 1px 5px;
  border-style: solid;
  border-color: #8080804d;
  border-color: ${props => props.$greaterthan3lessthan61 === "false" ? "red" : props.$focusorblur === "true" ? "#2196f3" : "#8080804d"};
  margin-bottom: 6px;
`

interface SpanNameProps {
  $hasvalueinputname: string;
}

export const SpanName = styled.span<SpanNameProps>`
  position: absolute;
  left: 16px;
  top: ${props => props.$hasvalueinputname === "true" ? "4px" : "13px"};
  font-size: ${props => props.$hasvalueinputname === "true" ? "11px" : "13px"};
  color: #a4a2a2;
  pointer-events: none;
  transition: top 0.2s ease, font-size 0.2s ease;
`

interface InputProps {
  $hasvalueinputname: string;
}

export const Input = styled.input<InputProps>`
  border: none;
  width: 100%;
  height: 100%;
  outline: none;
  padding: ${props => props.$hasvalueinputname === "true" ? "0px" : "13px"};
  padding-top: ${props => props.$hasvalueinputname === "true" && "18px"};
  padding-left: ${props => props.$hasvalueinputname === "true" && "17px"};
  padding-bottom: ${props => props.$hasvalueinputname === "true" && "8px"};
`

export const ContainerYourCpfWarning = styled.div`
  width: 25%;
`

interface ContainerCpfColorProps {
  $focusorblur: string;
  $greaterThan14Cpf: string;
}

export const ContainerCpfColor = styled.div<ContainerCpfColorProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  border-width: 1px 1px 1px 5px;
  border-style: solid;
  border-color: #8080804d;
  border-color: ${props => props.$greaterThan14Cpf === "false" ? "red" : props.$focusorblur === "true" ? "#2196f3" : "#8080804d"};
  margin-bottom: 6px;
 
`

interface SpanCpfProps {
  $hasvalueinputname: string;
}

export const SpanCpf = styled.span<SpanCpfProps>`
  position: absolute;
  left: 16px;
  top: 4px;
  font-size: 11px;
  color: #a4a2a2;
  pointer-events: none;
  transition: top 0.2s ease, font-size 0.2s ease;
`

interface InputCpfProps {
  $hasvalueinputname: string;
}

export const InputCpf = styled.input<InputCpfProps>`
  border: none;
  width: 100%;
  height: 100%;
  outline: none;
  padding: 0px;
  padding-top: 18px;
  padding-left: 17px;
  padding-bottom: 8px;
  user-select: none;
`