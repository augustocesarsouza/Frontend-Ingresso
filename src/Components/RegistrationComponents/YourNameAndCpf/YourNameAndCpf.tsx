import * as Styled from './styled';
import { useState, useEffect, useRef } from 'react';
import InputWarning from '../InputWarning/InputWarning';

interface YourNameAndCpfProps {
  verifyFields: boolean;
  setVerifyFields: React.Dispatch<React.SetStateAction<boolean>>;
  setValueOfName: React.Dispatch<React.SetStateAction<string>>;
  setValueOfCpf: React.Dispatch<React.SetStateAction<string>>;
}

const YourNameAndCpf = ({
  verifyFields,
  setVerifyFields,
  setValueOfName,
  setValueOfCpf,
}: YourNameAndCpfProps) => {
  const [numberYes, setNumberYes] = useState(false);
  const [backspacePress, setBackspacePress] = useState(false);
  const [valueNumber, setValueNumber] = useState('');
  const [FocusOrBlur, setFocusOrBlur] = useState(false);
  const [FocusOrBlurCpf, setFocusOrBlurCpf] = useState(false);
  const [hasValueInputName, setHasValueInputName] = useState(false);
  const [greaterThan3LessThan61, setGreaterThan3LessThan61] = useState(true);
  const [greaterThan14Cpf, setGreaterThan14Cpf] = useState(true);

  const refInputCpf = useRef<HTMLInputElement | null>(null);
  const inputNameRef = useRef<HTMLInputElement | null>(null);

  useEffect(() => {
    if (verifyFields) {
      const nameValue = inputNameRef.current.value;
      const cpfValue = refInputCpf.current.value;
      setValueOfName(nameValue);
      setValueOfCpf(cpfValue);

      if (nameValue.length < 3) {
        setVerifyFields(false);
        setGreaterThan3LessThan61(false);
      }

      let count = 0;
      for (let i = 0; i < cpfValue.length; i++) {
        if (
          refInputCpf.current.value[i] !== '.' &&
          refInputCpf.current.value[i] !== '-' &&
          refInputCpf.current.value[i] !== '_'
        ) {
          count++;
        }
      }

      if (count < 11) {
        setVerifyFields(false);
        setGreaterThan14Cpf(false);
      }
    }
  }, [inputNameRef.current, refInputCpf.current, verifyFields]);

  const handleChangeInputName = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.value.length > 0) {
      setHasValueInputName(true);

      if (e.target.value.length >= 3) {
        setGreaterThan3LessThan61(true);
      } else {
        setGreaterThan3LessThan61(false);
      }
    } else {
      setHasValueInputName(false);
      setGreaterThan3LessThan61(false);
    }
  };

  const handleCLick = (e: KeyboardEvent) => {
    if (e.key === 'Backspace') {
      setBackspacePress(true);
    } else {
      setBackspacePress(false);
    }

    if (!isNaN(Number(e.key))) {
      setValueNumber(e.key);
      setNumberYes(true);
    } else {
      setNumberYes(false);
    }
  };

  useEffect(() => {
    document.addEventListener('keydown', handleCLick);

    return () => {
      document.removeEventListener('keydown', handleCLick);
    };
  }, []);

  const [hasValueInputCpf, setHasValueInputCpf] = useState('___.___.___-__');
  const [cursorPositionIndex, setCursorPositionIndex] = useState(0);
  const [cursorPositionIndexDelete, setCursorPositionIndexDelete] = useState(0);

  const handleChangeInputCpf = (e: React.ChangeEvent<HTMLInputElement>) => {
    let count = 0;
    for (let i = 0; i < e.target.value.length; i++) {
      if (e.target.value[i] !== '.' && e.target.value[i] !== '-' && e.target.value[i] !== '_') {
        count++;
      }
    }

    if (count >= 11) {
      setGreaterThan14Cpf(true);
    } else {
      setGreaterThan14Cpf(false);
    }

    if (backspacePress) {
      const firstUnderscoreIndex = hasValueInputCpf.indexOf('_');

      if (firstUnderscoreIndex >= 0) {
        let positionSub = firstUnderscoreIndex - 1;
        if (hasValueInputCpf[positionSub] === '.' || hasValueInputCpf[positionSub] === '-') {
          positionSub = positionSub - 1;
        }

        if (positionSub >= 0) {
          const newValue =
            hasValueInputCpf.substring(0, positionSub) + // pega uma parte da string até onde quero deleta coloco no lugar _
            '_' +
            hasValueInputCpf.substring(positionSub + 1); // pego a parte depois do index que eu inseri _, pego resto e junto

          setHasValueInputCpf(newValue);
        }

        setCursorPositionIndexDelete(positionSub);
      }

      if (firstUnderscoreIndex === -1) {
        const newValue = hasValueInputCpf.substring(0, hasValueInputCpf.length - 1) + '_';

        setHasValueInputCpf(newValue);
        setCursorPositionIndexDelete(hasValueInputCpf.length - 1);
      }
    }

    if (numberYes) {
      const original = hasValueInputCpf;
      let replaced = false;
      let cursorPosition = null;

      const newValue = original.split('').map((char, index) => {
        if (char === '_' && !replaced) {
          replaced = true;
          cursorPosition = index + 1;
          return valueNumber;
        } else {
          return char;
        }
      });

      setHasValueInputCpf(newValue.join(''));
      if (cursorPosition !== null) {
        setCursorPositionIndex(cursorPosition);
      }
    }
  };

  useEffect(() => {
    if (cursorPositionIndex !== null) {
      if (refInputCpf.current) {
        refInputCpf.current.setSelectionRange(cursorPositionIndex, cursorPositionIndex);
      }
    }
  }, [cursorPositionIndex]);

  useEffect(() => {
    if (cursorPositionIndexDelete !== null) {
      if (refInputCpf.current) {
        refInputCpf.current.setSelectionRange(cursorPositionIndexDelete, cursorPositionIndexDelete);
      }
    }
  }, [cursorPositionIndexDelete]);

  const handleFocus = () => {
    setFocusOrBlur(true);
  };

  const handleBlur = () => {
    setFocusOrBlur(false);
  };

  const handleFocusCpf = () => {
    setFocusOrBlurCpf(true);
  };

  const handleBlurCpf = () => {
    setFocusOrBlurCpf(false);
  };

  const handleClickCpf = () => {
    if (refInputCpf.current) {
      const firstUnderscoreIndex = hasValueInputCpf.indexOf('_');

      refInputCpf.current.setSelectionRange(firstUnderscoreIndex, firstUnderscoreIndex);
    }
  };

  return (
    <Styled.ContainerInputExternal>
      <Styled.ContainerYourNameAndWarning>
        <Styled.ContainerAllColors
          $focusorblur={String(FocusOrBlur)}
          $greaterthan3lessthan61={String(greaterThan3LessThan61)}
        >
          <Styled.SpanName $hasvalueinputname={String(hasValueInputName)}>
            *Como quer ser chamado
          </Styled.SpanName>
          <Styled.Input
            type="text"
            id="input-your-name"
            ref={inputNameRef}
            onChange={handleChangeInputName}
            $hasvalueinputname={String(hasValueInputName)}
            onFocus={handleFocus}
            onBlur={handleBlur}
          />
        </Styled.ContainerAllColors>
        <InputWarning
          Children={'O nome precisa ter entre 3 e 60 caracteres e não possuir caracteres especiais'}
          enableWarning={greaterThan3LessThan61}
        />
      </Styled.ContainerYourNameAndWarning>
      <Styled.ContainerYourCpfWarning>
        <Styled.ContainerCpfColor
          $focusorblur={String(FocusOrBlurCpf)}
          $greaterThan14Cpf={String(greaterThan14Cpf)}
        >
          <Styled.SpanCpf $hasvalueinputname={String(hasValueInputName)}>*CPF</Styled.SpanCpf>
          <Styled.InputCpf
            type="text"
            maxLength={24}
            value={hasValueInputCpf}
            onChange={handleChangeInputCpf}
            $hasvalueinputname={String(hasValueInputName)}
            onFocus={handleFocusCpf}
            onBlur={handleBlurCpf}
            onClick={handleClickCpf}
            ref={refInputCpf}
            id="input-cpf"
          />
        </Styled.ContainerCpfColor>
        <InputWarning
          Children={'É obrigatório preencher o CPF.'}
          enableWarning={greaterThan14Cpf}
        />
      </Styled.ContainerYourCpfWarning>
    </Styled.ContainerInputExternal>
  );
};

export default YourNameAndCpf;
