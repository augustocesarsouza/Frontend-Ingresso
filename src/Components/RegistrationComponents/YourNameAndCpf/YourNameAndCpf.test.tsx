import { render, screen } from '@testing-library/react';
import YourNameAndCpf from './YourNameAndCpf';

describe('YourNameAndCpf', () => {
  test('should render input Your Name', () => {
    const { container } = render(<YourNameAndCpf />);

    const spanElement = container.querySelector('#input-your-name');
    expect(spanElement).toBeInTheDocument();
  });

  test('should render input Cpf', () => {
    const { container } = render(<YourNameAndCpf />);

    const spanElement = container.querySelector('#input-cpf');
    expect(spanElement).toBeInTheDocument();
  });

  test('should render span your name', () => {
    render(<YourNameAndCpf />);

    const spanElement = screen.getByText('*Como quer ser chamado');
    expect(spanElement).toBeInTheDocument();
  });

  test('should render span CPF', () => {
    render(<YourNameAndCpf />);

    const spanElement = screen.getByText('*CPF');
    expect(spanElement).toBeInTheDocument();
  });

  test('Snapshot', () => {
    const { container } = render(<YourNameAndCpf />);

    expect(container).toMatchSnapshot();
  });
});
