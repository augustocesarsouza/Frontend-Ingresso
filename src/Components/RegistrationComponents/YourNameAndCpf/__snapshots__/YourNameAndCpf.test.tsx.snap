// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YourNameAndCpf Snapshot 1`] = `
.c0 {
  display: flex;
  gap: 20px;
}

.c1 {
  display: flex;
  flex-direction: column;
  width: 55%;
}

.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  border-width: 1px 1px 1px 5px;
  border-style: solid;
  border-color: #8080804d;
  border-color: #8080804d;
  margin-bottom: 6px;
}

.c3 {
  position: absolute;
  left: 16px;
  top: 13px;
  font-size: 13px;
  color: #a4a2a2;
  pointer-events: none;
  transition: top 0.2s ease,font-size 0.2s ease;
}

.c4 {
  border: none;
  width: 100%;
  height: 100%;
  outline: none;
  padding: 13px;
}

.c5 {
  width: 25%;
}

.c6 {
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  border-width: 1px 1px 1px 5px;
  border-style: solid;
  border-color: #8080804d;
  border-color: #8080804d;
  margin-bottom: 6px;
}

.c7 {
  position: absolute;
  left: 16px;
  top: 4px;
  font-size: 11px;
  color: #a4a2a2;
  pointer-events: none;
  transition: top 0.2s ease,font-size 0.2s ease;
}

.c8 {
  border: none;
  width: 100%;
  height: 100%;
  outline: none;
  padding: 0px;
  padding-top: 18px;
  padding-left: 17px;
  padding-bottom: 8px;
  user-select: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <span
          class="c3"
        >
          *Como quer ser chamado
        </span>
        <input
          class="c4"
          id="input-your-name"
          type="text"
        />
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <span
          class="c7"
        >
          *CPF
        </span>
        <input
          class="c8"
          id="input-cpf"
          maxlength="24"
          type="text"
          value="___.___.___-__"
        />
      </div>
    </div>
  </div>
</div>
`;
